public/app/plugins/datasource/loki/datasource.ts

```diff
@@ -368,6 +378,47 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
         return { status: 'error', message: message };
       });
   }
+
+  async annotationQuery(options: AnnotationQueryRequest<{ expr: string }>): Promise<AnnotationEvent[]> {
+    const annotation = options.annotation;
+    const expr = annotation.expr || '';
+
+    if (!expr) {
+      return [];
+    }
+
+    const target: LokiQuery = {
+      refId: options.annotation.name,
+      expr,
+      live: false,
+    };
+
+    const query: DataQueryRequest<LokiQuery> = {
+      requestId: '',
+      timezone: 'utc',
+      range: {
+        from: dateTime(options.range.from),
+        to: dateTime(options.range.to),
+      },
+      targets: [target],
+    } as any;
```

Comment:
There are lots of stuff that does not seem to be needed here? Not sure but definitely not sure where to get them in this context so that's why this ugly force cast.

Reply:
requestId is important as cancellations are linked to it. Try using `annotation-<annotation-name>` as requestId.